File: src/common/py/data/entities/project/project_job.py
import time
import typing
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from .project import ProjectID
from ..connector import ConnectorInstanceID
from ..user import UserID

ProjectJobID = typing.Tuple[ProjectID, ConnectorInstanceID]


@dataclass_json
@dataclass(kw_only=True)
class ProjectJob:
    """
    A project job that is currently active.

    Attributes:
        user_id: The ID of the user the job belongs to.
        project_id: The project ID.
        connector_instance: The connector instance ID.
        timestamp: The starting time.
        progress: The total progress (0.0 - 1.0).
        message: The current activity message.
    """

    user_id: UserID
    project_id: ProjectID
    connector_instance: ConnectorInstanceID

    timestamp: float = field(default_factory=lambda: time.time())

    progress: float = 0.0
    message: str = ""

File: src/common/web/data/entities/project/ProjectJob.ts
import { type ConnectorInstanceID } from "../connector/ConnectorInstance";
import { type UserID } from "../user/User";
import { type ProjectID } from "./Project";

/**
 * A project job that is currently active.
 *
 * @param user_id - The ID of the user the job belongs to.
 * @param project_id - The project ID.
 * @param connector_instance - The connector instance ID.
 * @param timestamp - The starting time.
 * @param progress - The total progress (0.0 - 1.0).
 * @param message - The current activity message.
 */
export class ProjectJob {
    public readonly user_id: UserID;
    public readonly project_id: ProjectID;
    public readonly connector_instance: ConnectorInstanceID;

    public readonly timestamp: number;

    public readonly progress: number;
    public readonly message: string;

    constructor(userID: UserID, projectID: ProjectID, connectorInstance: ConnectorInstanceID, timestamp: number, progress: number = 0.0, message: string = "") {
        this.user_id = userID;
        this.project_id = projectID;
        this.connector_instance = connectorInstance;

        this.timestamp = timestamp;

        this.progress = progress;
        this.message = message;
    }
}
